MM-1539 Added the ability to reply to earlier message threads by starting a post with ^

hmhealey opened the pull request:
Some notes with this:
-The message thread to which you're replying is highlighted. Currently, the post's text is just bolded, but this is just a placeholder until we get the final style in.
-When determining which post to respond to, we only consider the first post in each message thread

crspeller commented on web/react/components/create_post.jsx:
```diff
@@ -112,6 +131,63 @@ module.exports = React.createClass({
+                // only dispatch an event if something changed
+                if (rootId != this.state.rootId) {
+                    // set the parent id to match the root id so that we're replying to the first post in the thread
+                    var parentId = rootId;
```
Why do we have both parentId and rootId if we are just setting them to be the same?

hmhealey replied:
I'm not sure exactly. When you reply through the sidebar, it also sets them to be the same. I'll try to figure out why that is.

crspeller commented on web/react/components/create_post.jsx:
```diff
@@ -112,6 +131,63 @@ module.exports = React.createClass({
+        // look to see if the message begins with any carets to indicate that it's a reply
+        var replyMatch = messageText.match(/^\^+/g);
+        if (replyMatch) {
+            // the number of carets indicates how many message threads back we're replying to
```
Should we keep state of how many `^` we find and check that here? That way we don't have to execute below if the number of `^` has not changed.

hmhealey replied:
That's a good point. I'm doing a bit of that by not sending the events out if the rootId doesn't change, but it's still reparsing the message each time if it does begin with carets. I need to think about how that'll work when someone else sends a message while you're typing one though since we might want it to switch what you're replying to when that happens.

hmhealey commented:
Updated so that we now keep track of the caret count so that we aren't constantly looking for which thread we're responding to as the user types. Also added some better handling for edge cases (like a user sending a message that only consists of a caret or responding to a message that was deleted).
